### PR TITLE
Fix broken LCTs reported in HLT_L1SingleMu25 [11_1_X]

### PR DESCRIPTION
**Pull request by dildick**
#### PR description:

An inefficiency was reported in HLT_L1SingleMu25 in CMSSW_11_2_0_pre2, which was not present in CMSSW_11_2_0_pre1 (https://its.cern.ch/jira/browse/CMSLITDPG-903). Investigation showed only ME1/2 LCTs (oddly enough) with invalid pattern numbers. This was fixed with https://github.com/cms-sw/cmssw/commit/9354dff643d672ed1d864559d757a53082355968. 

Further investigation revealed that the CLCT processor readout function was not checking if CLCTs were valid, and that GEM-CSC motherboards do not check if incoming pads are valid. The code was also missing quality control. I added a number of `checkValid` functions that check the properties for ALCT/CLCT/LCT at various stages of the algorithm. LogErrors are printed in case values go out of bounds.

#### PR validation:

Tested on 10k events of 
- `/RelValSingleMuFlatPt2To100/CMSSW_11_0_0-PU25ns_110X_mcRun4_realistic_v3_2026D49PU200-v1/GEN-SIM-DIGI-RAW`
- `/RelValSingleMuFlatPt2To100/CMSSW_11_0_0-110X_mcRun4_realistic_v2_2026D49noPU-v1/GEN-SIM-DIGI-RAW`
I did not see LogErrors of invalid stubs.

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

Backport of https://github.com/cms-sw/cmssw/pull/30909.

### CODE REVIEW

**Comment by civanch** [comment]
please test

**Comment by dildick** [comment]
Interesting. I justed this branch in CMSSW_11_1_X_2020-07-25-1100 on a 11_1_0_pre8 relval sample (in particular ```root://cmsxrootd-site.fnal.gov//store/relval/CMSSW_11_1_0_pre8/RelValZMM_14/GEN-SIM-DIGI-RAW/111X_mcRun3_2021_realistic_v4-v1/20000/E1B44039-321A-284E-91FB-97CD10F43A48.root```). 

The number of wiregroups the pretrigger loops on is constrained to [0,47]. Somehow the wiregroup number starts inflating past the bounds of the for-loop, ultimately causing a segfault in the anode processor pretrigger function.

```
...
%MSG-e CSCAnodeLCTProcessor:   CSCTriggerPrimitivesProducer:simCscTriggerPrimitiveDigis  26-Jul-2020 15:58:09 CDT Run: 1 Event: 754
CSCALCTDigi with invalid wire-group: 799; allowed [0, 48]
%MSG
...
%MSG-e CSCAnodeLCTProcessor:   CSCTriggerPrimitivesProducer:simCscTriggerPrimitiveDigis  26-Jul-2020 15:58:09 CDT Run: 1 Event: 754
CSCALCTDigi with invalid wire-group: 1242; allowed [0, 48]
%MSG
...
%MSG-e CSCAnodeLCTProcessor:   CSCTriggerPrimitivesProducer:simCscTriggerPrimitiveDigis  26-Jul-2020 15:58:09 CDT Run: 1 Event: 754
CSCALCTDigi with invalid wire-group: 1244; allowed [0, 48]
%MSG
...
```

Not sure what is causing this. The anode pretrigger function did not change recently. I also did not see such weird behavior in the corresponding 11_2_X branch (https://github.com/cms-sw/cmssw/pull/30909). In fact, I don't recall ever seeing this.

It seems that - at the very least - I need to add another explicit check that the wiregroup in the for-loop is valid.

**Comment by dildick** [comment]
I noticed that ```if (lct.getKeyWG() > max_wire)``` in ```checkValid``` should be tightened to ```if (lct.getKeyWG() >= max_wire)```. E.g. max_wire returns 48 for ME1/1, but wiregroup numbering starts from 0.

**Comment by silviodonato** [comment]
@rekovic ?

**Comment by ptcox** [comment]
Hi Sven,

As I am sure you're aware,  all offline geometry counts from 1, so the 
maximum  number of wiregroups in ME1/1 is 48 and counts 1-48.
If you want to make a statement like 'wiregroup numbering starts from 0' 
please make sure it is emphasized that this is in the trigger code only. 
Otherwise non-experts will get very confused.

Regards, Tim

**Comment by rekovic** [comment]
hi @dildick 
Regarding your comment https://github.com/cms-sw/cmssw/pull/30914#issuecomment-664040983.
Do you see the same on the HLT TDR samples, and does this fix the inefficiency we have reported in [1] ?

Why would you not see this in 11_2_X is surprising to me.  Is there anything in the geometry and counting that is different b/w 11_1_X and 11_2_X (@civanch @ptcox) ?

[1]
https://hypernews.cern.ch/HyperNews/CMS/get/L1TriggerUpgrades/385/2.html

**Comment by dildick** [comment]
@rekovic I'll check it now on an HLT TDR sample.

**Comment by dildick** [comment]
@rekovic I tested this on `/RelValSingleMuFlatPt2To100/CMSSW_11_0_0-110X_mcRun4_realistic_v2_2026D49noPU-v1/GEN-SIM-DIGI-RAW` Efficiencies in ME3/1 and ME4/1 should be OK now. Some of these commits need to be ported to the 11_2_X PR.

**Comment by fwyzard** [comment]
@cmsbuid, please test

File: L1Trigger/CSCTriggerPrimitives/interface/CSCMotherboard.h [hunk 1]
```diff
@@ -118,6 +118,9 @@ class CSCMotherboard : public CSCBaseboard {
   unsigned int highMultiplicityBits_;
   bool useHighMultiplicityBits_;
 
+  // Use the new patterns according to the comparator code format
+  bool use_run3_patterns_;
```
Comment (fwyzard) [inline L1Trigger/CSCTriggerPrimitives/interface/CSCMotherboard.h:122]: here ...

Comment (fwyzard) [inline L1Trigger/CSCTriggerPrimitives/interface/CSCMotherboard.h:122]: here ...

File: L1Trigger/CSCTriggerPrimitives/src/CSCCathodeLCTProcessor.cc [hunk 2]
```diff
@@ -905,7 +908,7 @@ bool CSCCathodeLCTProcessor::patternFinding(
                 hit_layer[this_layer] = true;
                 layers_hit++;  // determines number of layers hit
                 // add this strip in this layer to the pattern we are currently considering
-                hits_single_pattern[this_layer][strip_num] = this_strip;
+                hits_single_pattern[this_layer][strip_num] = this_strip - stagger[this_layer];
```
Comment (fwyzard) [inline L1Trigger/CSCTriggerPrimitives/src/CSCCathodeLCTProcessor.cc:911]: ... and here.

**Review by fwyzard — COMMENTED** [review]
Hi @dildick , in this PR there are - in addition to the changes from #30909 - a couple of extra changes backported from 11.2.x .
Can you confirm that they are intended ?

File: L1Trigger/CSCTriggerPrimitives/src/CSCCathodeLCTProcessor.cc (hunk 2 above)
Comment (dildick) [reply L1Trigger/CSCTriggerPrimitives/src/CSCCathodeLCTProcessor.cc:911]: Yes, the half-strip stagger was supposed to have been fixed in 11_1_X. But ultimately only made it into 11_2_X. See https://github.com/cms-sw/cmssw/pull/29830. This commit https://github.com/cms-sw/cmssw/pull/29830/commits/29c7161708e0bfbe8872e69e36c5dd8fcba2460b. The container `hits_single_pattern` keeps track of which comparator hits were ultimately used to build a CLCT, which you need when you want to match CLCTs to simtracks through compator digis.

File: L1Trigger/CSCTriggerPrimitives/interface/CSCMotherboard.h (hunk 1 above)
Comment (dildick) [reply L1Trigger/CSCTriggerPrimitives/interface/CSCMotherboard.h:122]: yes. This is intentional.

File: L1Trigger/CSCTriggerPrimitives/src/CSCCathodeLCTProcessor.cc (hunk 2 above)
Comment (fwyzard) [reply L1Trigger/CSCTriggerPrimitives/src/CSCCathodeLCTProcessor.cc:911]: OK, thanks.

**Comment by fwyzard** [comment]
urgent

**Comment by fwyzard** [comment]
backport #30909

**Comment by fwyzard** [comment]
hold

needs to be updated following #30909

**Comment by dildick** [comment]
A few LCT efficiency plots (not for publication) from `/RelValSingleMuFlatPt2To100/CMSSW_11_0_0-PU25ns_110X_mcRun4_realistic_v3_2026D49PU200-v1/GEN-SIM-DIGI-RAW`

Eff_CSCLCT_ME11
Eff_CSCLCT_ME12
Eff_CSCLCT_ME13
Eff_CSCLCT_ME21
Eff_CSCLCT_ME22
Eff_CSCLCT_ME31
Eff_CSCLCT_ME32
Eff_CSCLCT_ME41
Eff_CSCLCT_ME42

**Comment by dildick** [comment]
Efficiencies largely above 90%. Outliers are (1) ME1/2, which does not have upgraded processors or motherboard algorithms, (2) ME2/1 near |eta|~1.7.

**Comment by dildick** [comment]
@tahuang1991 Any idea why the GE2/1-ME2/1 algorithm would loose 20% efficiency near |eta|~1.7?

**Comment by rekovic** [comment]
please test

**Comment by fwyzard** [comment]
unhold

**Comment by fwyzard** [comment]
@dildick thank you very much for the plots

I admit I am not familiar with them... do you have them before the fix ?

**Comment by dildick** [comment]
Previously, the ME3/1 and ME4/1 LCT efficiencies (on a PU0 SingleMu relval sample) looked like this: 

Screen Shot 2020-07-31 at 2 24 28 AM
Screen Shot 2020-07-31 at 2 24 33 AM

**Comment by rekovic** [comment]
@dildick 
We had reports of TkMu efficiency of about 20% in the EndCap.    Your plots show about 20% impovement in TP efficiency, and that is from 70+ to 90+.   

Do you have any plots of change in CSC eta-phi coordinates before your work on the updating the algorithms and now ?
Those should not be affected, right ?

**Comment by silviodonato** [comment]
merge

**Pull request merged**